include/tree_walker.h

```diff
@@ -0,0 +1,134 @@
+ * [0]; x: [0,0]; 1: [0,1]; 2: [1]
+ */
+using UnorderedTermPairMap =
+    std::unordered_map<Term, std::pair<Term, std::vector<int>>>;
+
+/** \enum
+ * Walker_Continue : rebuild the current term and continue
+ * Walker_Skip     : skip this term and all subterms
+ * Walker_Abort    : abort visiting
+ */
+enum TreeWalkerStepResult
+{
+  TreeWalker_Continue = 0,
+  TreeWalker_Skip,
+  TreeWalker_Abort
+};
+
+/** \class
+ * TreeWalker class.
+ * To implement your own walker, inherit this class and override the
+ * visit_term method.
+ *
+ * If clear_cache is set, the cache is cleared between calls to visit
+ * otherwise, the cache persists
+ *
+ * The user can optionally pass a pointer to a cache. If that pointer
+ * is non-null, it will be used in place of the internal cache.
+ *
+ * Important Note: The term arguments should belong to the solver provided
+ */
+
+class TreeWalker
+{
+ public:
+  TreeWalker(const smt::SmtSolver & solver,
+             bool clear_cache,
+             smt::UnorderedTermPairMap * ext_cache = nullptr)
+      : solver_(solver), clear_cache_(clear_cache), ext_cache_(ext_cache){};
+
+  /** Visit a term and all its subterms in a post-order traversal
+   *  @param term the term to visit
+   *  @return the value of cache[term]
+   *    if it has been cached and returns pair of the term itself and the empty
+   * path otherwise
+   */
+  std::pair<smt::Term, std::vector<int>> visit(smt::Term & node);
+
+ protected:
+  /** Visit a single term.
+   *  Implement this method in a derived class to change the behavior
+   *  of the walker.
+   *  By default, builds up cache from term to term's first observed
+   *   occurrence in the formula, where an occurrence is represented
+   *   by a pair giving the topmost node (formula in which term appears)
+   *   & the path giving position in the formula of the first occurrence
+   *   of the term.
+   *  The path is represented as a vector of ints that serve as coordinates
+   *   for a given occurrence by giving a list of the child numbers to take
+   *   from the topmost node in the tree representation of the formula down
+   *   to the given occurrence in the formula. For example, an empty vector
+   *   indicates the topmost node and the vector [0,1] indicates the node
+   *   that is the 1st child of the 0th child of the topmost node going down.
+   *   Concretely, for the formula x+1=2, each term and its respective path is:
+   *     x+1=2: []
+   *     x+1: [0]
+   *     x: [0,0]
+   *     1: [0,1]
+   *     2: [1]
+   *  @param formula the term taken to visit in visit, the topmost node of the
+   * visited formula
+   *  @param term the term to visit in visit_term, a particular sub-term of the
+   * formula (some node in the tree representation of the full formula being
+   * traversed in visit)
+   *  @param path the path for the particular term in formula, which we are
+   * visiting
+   *  @return a WalkerStepResult to tell the visit method how to proceed
```

Comment:
This should be `TreeWalkerStepResult` in the comment, unless you unify `TreeWalkerStepResult` and `WalkerStepResult`.